File: src/ingest/chunker.ts
/** Break-point-scored markdown chunking (DESIGN.md §7.2).
 *
 * Greedy fill to `targetTokens` while preferring structural break points
 * (headings, fences, rules, speaker turns, paragraph starts). Code fences are
 * never split: a hard cut that would land inside an open fence is extended
 * past the closing fence. Overlap is carried across arbitrary (hard) cuts,
 * but not across clean structural breaks — a heading always starts a fresh
 * chunk.
 *
 * Ported from `legacy/python/src/qkb/ingest/chunker.py`. */

import type { Chunk } from "../types.js";

const HEADING = /^(#{1,6})\s/;
const FENCE = /^(```|~~~)/;
const HR = /^(-{3,}|\*{3,})\s*$/;
const SPEAKER = /^\[\d{2}:\d{2}(:\d{2})?\]\s*\S/;
const LIST = /^(\s*)([-*+]|\d+\.)\s/;

const HEADING_SCORE: Record<number, number> = { 1: 100, 2: 90, 3: 80, 4: 70, 5: 60, 6: 50 };

export function estimateTokens(text: string): number {
  return text ? Math.max(1, Math.floor(text.length / 4)) : 0;
}

/** Score the break BEFORE each line. Breaks inside code fences score 0 (forbidden). */
function scoreLines(lines: string[]): number[] {
  const scores = new Array<number>(lines.length).fill(0);
  let inFence = false;
  let prevBlank = false;
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i] ?? "";
    if (FENCE.test(line)) {
      // an opening fence is a legal boundary; a closing fence is not
      scores[i] = inFence ? 0 : 80;
      inFence = !inFence;
      prevBlank = false;
      continue;
    }
    if (inFence) {
      scores[i] = 0;
      prevBlank = false;
      continue;
    }
    const m = HEADING.exec(line);
    if (m) {
      scores[i] = HEADING_SCORE[(m[1] ?? "").length] ?? 0;
    } else if (HR.test(line)) {
      scores[i] = 60;
    } else if (SPEAKER.test(line)) {
      scores[i] = 30;
    } else if (prevBlank && line.trim()) {
      scores[i] = 20;
    } else if (LIST.test(line)) {
      scores[i] = 5;
    } else if (line.trim()) {
      scores[i] = 1;
    }
    prevBlank = !line.trim();
  }
  return scores;
}

/** True where cutting BEFORE this line would split an open code fence. */
function fenceInterior(lines: string[]): boolean[] {
  const interior = new Array<boolean>(lines.length).fill(false);
  let inFence = false;
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i] ?? "";
    if (FENCE.test(line)) {
      if (inFence) {
        interior[i] = true; // closing fence must stay with its opener
      }
      inFence = !inFence;
    } else {
      interior[i] = inFence;
    }
  }
  return interior;
}

export function chunkText(text: string, targetTokens = 500, overlapPercent = 15): Chunk[] {
  if (!text.trim()) {
    return [];
  }
  const lines = text.split("\n");
  const scores = scoreLines(lines);
  const interior = fenceInterior(lines);
  const lineTokens = lines.map((line) => estimateTokens(line) + 1); // +1 for the newline
  const n = lines.length;

  const chunks: Chunk[] = [];
  const overlapTokens = Math.floor((targetTokens * overlapPercent) / 100);
  const minSize = Math.floor(targetTokens / 2);
  let start = 0;
  while (start < n) {
    let size = 0;
    let j = start;
    let bestCut = -1;
    let bestScore = 0.0;
    while (j < n) {
      size += lineTokens[j] ?? 0;
      j += 1;
      // `j` is now a candidate cut point (chunk = lines[start:j]); scores[j] rates
      // starting a new chunk at line j. Prefer the strongest structural break once
      // we have accumulated at least half a chunk.
      const scoreAtJ = scores[j] ?? 0;
      if (j < n && scoreAtJ > 0 && !interior[j] && size >= minSize) {
        if (scoreAtJ >= bestScore) {
          bestScore = scoreAtJ;
          bestCut = j;
        }
      }
      if (size >= targetTokens) {
        break;
      }
    }

    let structural = false;
    let cut: number;
    if (size < targetTokens) {
      cut = n; // reached EOF without filling a whole chunk
    } else if (bestCut > start) {
      cut = bestCut; // clean structural break
      structural = true;
    } else {
      cut = j; // hard cut mid-content...
      while (cut < n && interior[cut]) {
        // ...but never inside a code fence
        cut += 1;
      }
    }

    const body = lines
      .slice(start, cut)
      .join("\n")
      .replace(/^\n+|\n+$/g, "");
    if (body.trim()) {
      chunks.push({ index: chunks.length, text: body, tokenCount: estimateTokens(body) });
    }
    if (cut >= n) {
      break;
    }

    if (structural) {
      start = cut; // a heading/rule starts its own chunk — no overlap across it
      continue;
    }
    // carry overlap from the tail of this chunk into the next one
    let back = cut;
    let dist = 0;
    while (back > start && dist < overlapTokens) {
      dist += lineTokens[back - 1] ?? 0;
      back -= 1;
    }
    start = Math.max(back, start + 1);
    while (start < cut && interior[start]) {
      // never resume inside a fence
      start += 1;
    }
    if (start >= cut) {
      start = cut;
    }
  }
  return chunks;
}

File: legacy/python/src/qkb/ingest/chunker.py
"""Break-point-scored markdown chunking (DESIGN.md §7.2).

Greedy fill to ``target_tokens`` while preferring structural break points
(headings, fences, rules, speaker turns, paragraph starts). Code fences are
never split: a hard cut that would land inside an open fence is extended past
the closing fence. Overlap is carried across arbitrary (hard) cuts, but not
across clean structural breaks — a heading always starts a fresh chunk.
"""

from __future__ import annotations

import re

from qkb.models import Chunk

_HEADING = re.compile(r"^(#{1,6})\s")
_FENCE = re.compile(r"^(```|~~~)")
_HR = re.compile(r"^(-{3,}|\*{3,})\s*$")
_SPEAKER = re.compile(r"^\[\d{2}:\d{2}(:\d{2})?\]\s*\S")
_LIST = re.compile(r"^(\s*)([-*+]|\d+\.)\s")


def estimate_tokens(text: str) -> int:
    return max(1, len(text) // 4) if text else 0


def _score_lines(lines: list[str]) -> list[int]:
    """Score the break BEFORE each line. Breaks inside code fences score 0 (forbidden)."""
    scores = [0] * len(lines)
    in_fence = False
    prev_blank = False
    for i, line in enumerate(lines):
        if _FENCE.match(line):
            # an opening fence is a legal boundary; a closing fence is not
            scores[i] = 0 if in_fence else 80
            in_fence = not in_fence
            prev_blank = False
            continue
        if in_fence:
            scores[i] = 0
            prev_blank = False
            continue
        m = _HEADING.match(line)
        if m:
            scores[i] = {1: 100, 2: 90, 3: 80, 4: 70, 5: 60, 6: 50}[len(m.group(1))]
        elif _HR.match(line):
            scores[i] = 60
        elif _SPEAKER.match(line):
            scores[i] = 30
        elif prev_blank and line.strip():
            scores[i] = 20
        elif _LIST.match(line):
            scores[i] = 5
        elif line.strip():
            scores[i] = 1
        prev_blank = not line.strip()
    return scores


def _fence_interior(lines: list[str]) -> list[bool]:
    """True where cutting BEFORE this line would split an open code fence."""
    interior = [False] * len(lines)
    in_fence = False
    for i, line in enumerate(lines):
        if _FENCE.match(line):
            if in_fence:
                interior[i] = True  # closing fence must stay with its opener
            in_fence = not in_fence
        else:
            interior[i] = in_fence
    return interior


def chunk_text(text: str, target_tokens: int = 500, overlap_percent: int = 15) -> list[Chunk]:
    if not text.strip():
        return []
    lines = text.split("\n")
    scores = _score_lines(lines)
    interior = _fence_interior(lines)
    line_tokens = [estimate_tokens(line) + 1 for line in lines]  # +1 for the newline
    n = len(lines)

    chunks: list[Chunk] = []
    overlap_tokens = target_tokens * overlap_percent // 100
    min_size = target_tokens // 2
    start = 0
    while start < n:
        size = 0
        j = start
        best_cut = -1
        best_score = 0.0
        while j < n:
            size += line_tokens[j]
            j += 1
            # `j` is now a candidate cut point (chunk = lines[start:j]); scores[j] rates
            # starting a new chunk at line j. Prefer the strongest structural break once
            # we have accumulated at least half a chunk.
            if j < n and scores[j] > 0 and not interior[j] and size >= min_size:
                if scores[j] >= best_score:
                    best_score, best_cut = float(scores[j]), j
            if size >= target_tokens:
                break

        structural = False
        if size < target_tokens:
            cut = n  # reached EOF without filling a whole chunk
        elif best_cut > start:
            cut = best_cut  # clean structural break
            structural = True
        else:
            cut = j  # hard cut mid-content...
            while cut < n and interior[cut]:  # ...but never inside a code fence
                cut += 1

        body = "\n".join(lines[start:cut]).strip("\n")
        if body.strip():
            chunks.append(Chunk(index=len(chunks), text=body, token_count=estimate_tokens(body)))
        if cut >= n:
            break

        if structural:
            start = cut  # a heading/rule starts its own chunk — no overlap across it
            continue
        # carry overlap from the tail of this chunk into the next one
        back, dist = cut, 0
        while back > start and dist < overlap_tokens:
            dist += line_tokens[back - 1]
            back -= 1
        start = max(back, start + 1)
        while start < cut and interior[start]:  # never resume inside a fence
            start += 1
        if start >= cut:
            start = cut
    return chunks
